### PR TITLE
Gc threadpool

### PR DESCRIPTION
**Pull request by ashtul**
replace thread-per-index

### CODE REVIEW

**Comment by ashtul** [comment]
> @ashtul you can use union with 3 types: SCHEDULE, BG_FORCE_INVOKED, FORCE_INVOKED or some other names you chose

I can't because I use the pointer to release the task after using it. I can use `1` for BG_FORCE_INVOKED instead of `DEADBEEF`

**Comment by ashtul** [comment]
fix #524

File: src/gc.c [hunk 1]
```diff
@@ -4,35 +4,26 @@
...
-  RedisModuleBlockedClient* ret = NULL;
-  DLLIST_node* nn = dllist_pop_tail(&ctx->clients);
-  if (nn) {
-    BlockClient* bc = DLLIST_ITEM(nn, BlockClient, llnode);
-    ret = bc->bClient;
-    rm_free(bc);
-  }
-  pthread_mutex_unlock(&ctx->lock);
-  return ret;
+#define DEADBEEF (void*)0xBADC0FFEE0DDF00D
```
Comment (MeirShpilraien) [inline src/gc.c:13]: Didn't we say we will do it without this DEADBEEF?

File: src/gc.c [hunk 2]
```diff
@@ -62,44 +51,116 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+  thpool_add_work(gcThreadpool_g, threadCallback, data);
 }
 
 void GCContext_Start(GCContext* gc) {
-  struct timespec interval = gc->callbacks.getInterval(gc->gcCtx);
-  gc->timer = RMUtil_NewPeriodicTimer(GCContext_PeriodicCallback, GCContext_OnTerm, gc, interval);
+  GCTaskCtx* task = GCTaskCreate(gc, NULL);
+  gc->timerID = scheduleNext(task);
+  if (gc->timerID == 0) {
+    rm_free(task);
```
Comment (MeirShpilraien) [inline src/gc.c:142]: Log?

File: src/gc.c [hunk 3]
```diff
@@ -112,11 +173,34 @@ void GCContext_OnDelete(GCContext* gc) {
   }
 }
 
+void GCContext_CommonForceInvoke(GCContext* gc, RedisModuleBlockedClient* bc) {
+  if (gc->stopped) {
+    return;
```
Comment (MeirShpilraien) [inline src/gc.c:178]: Log?

File: src/gc.c [hunk 4]
```diff
@@ -112,11 +173,34 @@ void GCContext_OnDelete(GCContext* gc) {
...
+void GC_ThreadPoolStart() {
+  if (gcThreadpool_g == NULL) {
+    gcThreadpool_g = thpool_init(1);
+  }
+}
+
+void GC_ThreadPoolDestroy() {
+  if (gcThreadpool_g != NULL) {
+    RedisModule_ThreadSafeContextUnlock(RSDummyContext);
+    thpool_destroy(gcThreadpool_g);
```
Comment (MeirShpilraien) [inline src/gc.c:202]: I think the thpool_destroy are waiting for the last task to finish but you acquire the lock so the task will not be able to finish, deadlock.

File: src/gc.c (hunk 1 above)
Comment (ashtul) [reply src/gc.c:13]: Just trim for 32bit

File: src/gc.c (hunk 2 above)
Comment (ashtul) [reply src/gc.c:142]: This can only happen with debug command. 
RedisModule_CreateTimer always returns a value

File: src/gc.c (hunk 4 above)
Comment (ashtul) [reply src/gc.c:202]: We had a deadlock which is why we use `Unlock`

File: src/gc.c [hunk 5]
```diff
@@ -175,10 +176,11 @@ void GCContext_OnDelete(GCContext* gc) {
 
 void GCContext_CommonForceInvoke(GCContext* gc, RedisModuleBlockedClient* bc) {
   if (gc->stopped) {
+    RedisModule_Log(RSDummyContext, "warning", "ForceInvokeGC command received after shut down");
```
Comment (MeirShpilraien) [inline src/gc.c:179]: :+1:

**Comment by ashtul** [comment]
@MeirShpilraien @mnunberg 
Added env var `RS_GLOBAL_DTORS` to CircleCi and call `getEnv` before running destructors.

**Pull request merged**